src/lib.rs

```diff
@@ -179,94 +177,52 @@ impl<T: AsRef<str>> Grid<T> {
-        // needed to display all cells is 1.
-        1
-    }
-
-    fn width_dimensions(&self, maximum_width: usize) -> Option<Dimensions> {
-        if self.widest_cell_width > maximum_width {
-            // Largest cell is wider than maximum width; it is impossible to fit.
-            return None;
-        }
-
-        if self.cells.is_empty() {
-            return Some(Dimensions {
-                num_lines: 0,
-                widths: Vec::new(),
-            });
-        }
-
+    fn width_dimensions(&mut self) -> Dimensions {
         if self.cells.len() == 1 {
             let cell_widths = self.widths[0];
-            return Some(Dimensions {
+            return Dimensions {
                 num_lines: 1,
                 widths: vec![cell_widths],
-            });
+            };
         }
 
-        let theoretical_max_num_lines = self.theoretical_max_num_lines(maximum_width);
-        if theoretical_max_num_lines == 1 {
-            // This if—statement is necessary for the function to work correctly
-            // for small inputs.
-            return Some(Dimensions {
-                num_lines: 1,
-                widths: self.widths.clone(),
-            });
+        // Calculate widest column size with separator.
+        let widest_column = self.widest_cell_width + self.options.filling.width();
+        // If it exceeds terminal's width, return, since it is impossible to fit.
+        if widest_column > self.options.width {
+            return Dimensions {
+                num_lines: self.cells.len(),
+                widths: vec![self.widest_cell_width],
+            };
         }
-        // Instead of numbers of columns, try to find the fewest number of *lines*
-        // that the output will fit in.
-        let mut smallest_dimensions_yet = None;
-        for num_lines in (1..=theoretical_max_num_lines).rev() {
-            // The number of columns is the number of cells divided by the number
-            // of lines, *rounded up*.
-            let num_columns = div_ceil(self.cells.len(), num_lines);
-
-            // Early abort: if there are so many columns that the width of the
-            // *column separators* is bigger than the width of the screen, then
-            // don’t even try to tabulate it.
-            // This is actually a necessary check, because the width is stored as
-            // a usize, and making it go negative makes it huge instead, but it
-            // also serves as a speed-up.
-            let total_separator_width = (num_columns - 1) * self.options.filling.width();
-            if maximum_width < total_separator_width {
-                continue;
-            }
 
-            // Remove the separator width from the available space.
-            let adjusted_width = maximum_width - total_separator_width;
+        // Calculate number of columns with widest column size.
+        let max_num_columns = self.options.width / widest_column;
```

Comment:
This confused me for a long time, because I took the variable name at face value, but it's actually more the _minimum_ number of columns, or rather, if you `min` this with the number of cells it's the minimum. Like this:
```rust
        let min_columns = self
            .cells
            .len()
            .min((self.options.width + self.options.filling.width()) / widest_column);
```
I think that removes the need for the `appr_num_columns` calculation below.